### ISSUE TITLE
Schema validation modifies exported config object in .eslintrc.js

### ISSUE THREAD

**Issue by andreineculau** [eslint/eslint#12592]
**Tell us about your environment**

* **ESLint Version:** 6.7.0
* **Node Version:** 12.13.1
* **npm Version:** 6.12.1

**What parser (default, Babel-ESLint, etc.) are you using?** babel-eslint

**Please show your full configuration:**

<details>
<summary>Configuration</summary>

```js
// .eslintrc.basic.js
module.exports = {
  plugins: [],

  extends: [
    'eslint:all'
  ],

  parser: 'espree',

  parserOptions: {
    ecmaVersion: 7,
    sourceType: 'module'
  },

  settings: {},

  env: {
    es6: true
  },

  rules: {
    'camelcase': ['warn', {properties: 'never'}]
  }
}

// .eslintrc.babel.js
let _basic = require('./.eslintrc.basic');

module.exports = {
  parser: 'babel-eslint',

  parserOptions: {
    sourceType: 'module'
  },

  plugins: [
    'babel'
  ],

  rules: {
    camelcase: 'off',
    'babel/camelcase': _basic.rules.camelcase
  }
};

// .eslintrc.js
module.exports: {
  extends: [
    './.eslintrc.basic.js',
    './.eslintrc.babel.js'
  ]
}
```

</details>

**What did you do? Please include the actual source code causing the issue, as well as the command that you used to run ESLint.**

```js
N/A
```

```bash
N/A
```

**What did you expect to happen?**

I expected no errors.

**What actually happened? Please include the actual, raw output from ESLint.**

```
	Configuration for rule "babel/camelcase" is invalid:
	Value {"properties":"never","ignoreDestructuring":false,"ignoreImports":false} should NOT have additional properties.
```

This happens because of https://github.com/eslint/eslint/pull/12528 adding `ignoreImports` with default value `failse`, and because the defaults of `camelcase` are merged into the same object in `.eslintrc.basic.js:rules.camelcase`, which is also the same reference in `.eslintrc.babel.js:rules.@babel/camelcase`.

**Are you willing to submit a pull request to fix this bug?**

Sure.

**Comment by kaicataldo** [comment]
`eslint-plugin-babel` is maintained by a different team. Please file an issue with https://github.com/babel/eslint-plugin-babel.

**Comment by andreineculau** [comment]
This has nothing to do with babel or any other plugin. I'm on the mobile now but i think the problem can be reproduced via

let errorConfig = ['error'];
module.exports = {
  rules: {camelcase: errorConfig, 'default-case': errorConfig}
};

Where eslint will complain that default-case doesn't know of camelcase's options with default values.

**Comment by andreineculau** [comment]
correction: let errorConfig = ['error', {}]

**Comment by ljharb** [comment]
I would expect the json schema to reject an empty object, and to require at least one key to be present. Is that the case here?

**Comment by andreineculau** [comment]
Eslint barfs with
```
	Configuration for rule "default-case" is invalid:
	Value {"ignoreDestructuring":false} should NOT have additional properties.
```

due to configuring ajv with `useDefaults: true` in https://github.com/eslint/eslint/blob/219aecb78bc646d44bad27dc775a9b3d3dc58232/lib/shared/ajv.js#L21, which [means](https://github.com/epoberezkin/ajv#assigning-defaults) that the options for `camelcase` will mutate into `{ignoreDestructuring: false}`.

I don't think anyone expects the configuration passed to `module.exports` in a `.eslintrc.js` file to be mutated. That is why I went ahead and said you eslint should dereference the configuration e.g. `config = JSON.parse(JSON.stringify(config)); // do whatever you want with config from here on`

My no-plugins example is coerced and doesn't make sense. I don't know why would you want to share the same options to two rules. BUT, now that I have shown you that the behavior is not related to plugins, but to eslint itself, the pattern of sharing configs "upstream" to the babel or typescript plugin is useful because many of their rules are supersets of the built-in ones and it allows you maintain the same configuration.

<details>
Real life scenario: babel has a `babel/camelcase` rule, and typescript has a `typescript/camelcase` one, both supersets of the built-in `camelcase` rule. And the recommendation goes like:

```
{
  camelcase: 'off', // turn off the built-in rule
  'babel/camelcase': ['error', {ignoreDestructuring: true}] // activate the superset rule
}
```

but if you maintain a `.eslintrc.basic.js` and a `.eslintrc.babel.js` meant just for your babel projects, then you need to

* either sync the configuration (i.e. if you want `warn` instead of `error`, then make sure to remember to change it in both files)
* or change `.eslintrc.babel.js` to smth like `let basic = require('./eslintrc.basic'); modules.exports = {rules: {camelcase: 'off', 'babel/camelcase': basic.rules.camelcase}};`.
</details>

**Comment by ljharb** [comment]
ah, gotcha. your use of the term "dereference" confused me since JS has no dereferencing and no pointers; objects are "reference values" :-)

So specifically, you're saying that the validator should apply to the original config, not to the ajv-mutated one?

**Comment by andreineculau** [comment]
yup

**Comment by ljharb** [comment]
This seems important since if unfixed, it leaves inaccurate error messages.

**Comment by kaicataldo** [comment]
Agreed. Reopened.

**Comment by platinumazure** [comment]
Could we retitle this, please?

**Comment by platinumazure** [comment]
I've edited the title to describe what I understand the issue to be. Please let me know if I've misunderstood.

I would like to apologize for my previous comment, which might have come off as curt and rude. I was typing while multitasking and I should have waited until I could give my full attention and communicate in the way our values community members deserve. I failed to do that, and for that I apologize. I have removed the original comment.

**Comment by kaicataldo** [comment]
@eslint/eslint-team How do we want to proceed here?

**Comment by g7tro4w** [comment]
Any news?

**Comment by nzakas** [comment]
@kaicataldo can you summarize the issue and how you’d like to proceed? (It looks like you’re the assignee.)

**Comment by kaicataldo** [comment]
Sure thing. As I understand it, the issue here is that, because we use the [`useDefaults`](https://github.com/eslint/eslint/blob/219aecb78bc646d44bad27dc775a9b3d3dc58232/lib/shared/ajv.js#L21) option when we validate with `ajv`, the defaults get merged into the config during validation. Since this mutates the config it's given, if that config is referenced elsewhere (such as in multiple config files like the issue author's example has), it can end up having unintended consequences.

I don't see any downsides to cloning the config before validation - thoughts from the rest of the team? Note that this will also be relevant with the [new config file format](https://github.com/eslint/rfcs/pull/9) we're planning to move towards.

**Comment by nzakas** [comment]
Makes sense. No objections to cloning from me.

**Comment by mdjermanovic** [comment]
> I don't see any downsides to cloning the config before validation - thoughts from the rest of the team?

Sounds good, assuming that the rule will get the clone (which might have been modified during the validation), not the original config.

**Comment by nzakas** [comment]
Just for some context, the Ajv docs make it clear that `useDefaults` modifies original data: https://github.com/ajv-validator/ajv#assigning-defaults

@epoberezkin we seem to be having an issue with the behavior of `useDefaults`, as we missed that it modifies the object that is passed. Would it be at all feasible for Ajv to have an option to clone the object instead of modifying it? It seems like a more common case to want the original object to remain intact.

**Comment by epoberezkin** [comment]
@nzakas my decision not to clone in Ajv was because in general case you cannot clone a JavaScript object. You can clone plain objects (e.g. the result of JSON.parse) and quite a few other things, but ajv can be used to validate all types of objects... Doing it in eslint would be easier because eslint can rely on some assumptions about what the object is - whether `JSON.parse(JSON.stringify(obj))` is good enough or some deep clone is needed that would correctly clone some other things and not only plain objects.

It seems that these two places need to be changed:

https://github.com/eslint/eslint/blob/33efd71d7c3496b4b9cbfe006280527064940826/lib/shared/config-validator.js#L96

https://github.com/eslint/eslint/blob/33efd71d7c3496b4b9cbfe006280527064940826/lib/shared/config-validator.js#L258

where `ajv.compile(schema)` would be replaced with say `compile(schema)` where `compile` is something like:

```javascript
function compile(schema) {
  const validate = ajv.compile(schema);
  return function v(data) {
  }
}

**Comment by epoberezkin** [comment]
@nzakas @kaicataldo let me know what you think - happy to help if needed.

**Comment by nzakas** [comment]
Yeah, that’s what we are trying to do in https://github.com/eslint/eslint/pull/13034

Turns out to be tricky even with assumptions.

**Comment by kaicataldo** [comment]
Reopening this so we can find another solution.

**Comment by btmills** [comment]
This was auto-closed when the revert commit was merged. Re-opening since it’s still outstanding.

**Comment by nzakas** [comment]
I'll tackle this as part of the new config system.

**Comment by Slapbox** [comment]
Holy cow I didn't expect tweaking our default-case `commentPattern` option to send me down such rabbit hole! Not complaining, just surprised.

Any news on this? Anyone know a workaround to get this config working?

```js
    'default-case': ['error', { commentPattern: /^no-default-case$/i }]
```

**Comment by mdjermanovic** [comment]
> Any news on this? Anyone know a workaround to get this config working?
> 
> ```js
>     'default-case': ['error', { commentPattern: /^no-default-case$/i }]
> ```

@Slapbox this isn't related to the schema validation issue, `default-case` rule by design expects a string value as `commentPattern`:

```js
'default-case': ['error', { commentPattern: "^no-default-case$" }]
```

It doesn't support specifying flags. If you want to allow uppercase in all letters, you can use this workaround:

```js
'default-case': ['error', { commentPattern: "^[nN][oO]-[dD][eE][fF][aA][uU][lL][tT]-[cC][aA][sS][eE]$" }]
```

[online demo](https://eslint.org/demo#eyJ0ZXh0IjoiLyogZXNsaW50IGRlZmF1bHQtY2FzZTogWydlcnJvcicsIHsgY29tbWVudFBhdHRlcm46IFwiXltuTl1bb09dLVtkRF1bZUVdW2ZGXVthQV1bdVVdW2xMXVt0VF0tW2NDXVthQV1bc1NdW2VFXSRcIiB9XSAqL1xuXG5zd2l0Y2ggKGZvbykge1xuICAgIGNhc2UgYmFyOlxuICAgICAgIGJheigpO1xuICAgICAgIGJyZWFrO1xuICAgIFxuICAgIC8vIG5vLWRlZmF1bHQtY2FzZVxufVxuXG5zd2l0Y2ggKGZvbykge1xuICAgIGNhc2UgYmFyOlxuICAgICAgIGJheigpO1xuICAgICAgIGJyZWFrO1xuICAgIFxuICAgIC8vIE5vLURlRmF1TFQtQ2FTZVxufSIsIm9wdGlvbnMiOnsicGFyc2VyT3B0aW9ucyI6eyJlY21hVmVyc2lvbiI6MTIsInNvdXJjZVR5cGUiOiJzY3JpcHQiLCJlY21hRmVhdHVyZXMiOnt9fSwicnVsZXMiOnt9LCJlbnYiOnt9fX0=)

**Comment by Slapbox** [comment]
Derp. This was the closest issue I could find at the end of the day with a burned out brain. Big thanks @mdjermanovic!
